pandas/core/dtypes/cast.py

```diff
@@ -1562,26 +1565,28 @@ def maybe_cast_to_datetime(value, dtype: Optional[DtypeObj]):
 
                     try:
                         if is_datetime64:
```

Comment:
@simonjayhawkins im getting a bunch of mypy complaints locally that IIUC should be fixed by doing `value = cast(np.ndarray, value)` after L1556, but that isnt improving it.  neither is `assert isinstance(value, np.ndarray)`.  thoughts?

Reply:
it's a typevar issue. the return type is not necessarily the same as value?

changing
```diff
diff --git a/pandas/core/dtypes/cast.py b/pandas/core/dtypes/cast.py
index ab5cbdd823..6abbc67454 100644
--- a/pandas/core/dtypes/cast.py
+++ b/pandas/core/dtypes/cast.py
@@ -1500,8 +1500,8 @@ def maybe_infer_to_datetimelike(


 def maybe_cast_to_datetime(
-    value: Union[ArrayLike, list], dtype: Optional[DtypeObj]
-) -> Union[ArrayLike, list]:
+    value: Union[ExtensionArray, np.ndarray, list], dtype: Optional[DtypeObj]
+) -> Union[ExtensionArray, np.ndarray, list]:
     """
     try to cast the array/value to a datetimelike dtype, converting float
     nan to iNaT
```

just leaves
```
pandas\core\dtypes\cast.py:1602: error: Item "List[Any]" of "Union[ExtensionArray, Any, List[Any]]" has no attribute "dtype" 
 [union-attr]
pandas\core\dtypes\cast.py:1603: error: Item "List[Any]" of "Union[ExtensionArray, Any, List[Any]]" has no attribute "astype"  [union-attr]
pandas\core\dtypes\cast.py:1616: error: Item "List[Any]" of "Union[ExtensionArray, Any, List[Any]]" has no attribute "dtype" 
 [union-attr]
pandas\core\dtypes\cast.py:1622: error: Item "List[Any]" of "Union[ExtensionArray, Any, List[Any]]" has no attribute "dtype" 
 [union-attr]
```

after/if #36092 gets merged, the first thing will be to change the typevar to a union in pandas._typing to get rid of a couple hundred ignores. https://github.com/simonjayhawkins/pandas/compare/numpy-types...simonjayhawkins:numpy-fixes

https://github.com/pandas-dev/pandas/pull/36092#issuecomment-724696623

Reply:
that fixed it, thanks